### PR TITLE
TST: Mark slow tests in astropy/samp

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
A quick follow up to #16064
I found the 10 longest (non-slow, not parametrized) tests with `pytest astropy/ --timer-top-n 20` (with `pytest-timer`), which is also about the same set of tests that take
longer than a second on my machine (supposedly a pretty fast M2).
On my install, `pytest astropy` takes about 2'15 on main, and 1'55 with this branch, hence I claim that skipping those 10 tests (out of 24k) might save about 10% CI time. 

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
I'll have to ponder later if the cost of not running this for all combo is worth the benefit. Not like we just sit staring at the CI until it finishes.

**Comment by mhvk** [comment]
I like this idea, but it would be good to get some input on whether these specific tests are things that are likely to fail on just one architecture.

Also, some of the slowest tests may be ones that are excessively parametrized (or the `hypothesis` ones in `astropy.time.tests.test_precision`). I think there may be a way to look at those - maybe https://github.com/numpy/numpy/issues/25472#issuecomment-1868320220? But obviously fine to do that later.

**Review by pllim — DISMISSED** [review]
In general, I am okay with marking SAMP as slow, but not comfortable with `io.fits` and `utils`. FITS is a can of worms and `utils` is backbone of many things. I want them both to be tested widely. A bit of slowness in CI is a price I am willing to pay (not that we're paying, har har).

Thank you for your understanding!

File: astropy/io/fits/hdu/compressed/tests/test_compressed.py [hunk 1]
```diff
@@ -204,6 +204,7 @@ def test_disable_image_compression(self):
         with fits.open(self.data("comp.fits")) as hdul:
             assert isinstance(hdul[1], fits.CompImageHDU)
 
+    @pytest.mark.slow
```
Comment (mhvk) [inline astropy/io/fits/hdu/compressed/tests/test_compressed.py:207]: Hmm, looking at the actual test, I think the only reason it is slow is because it *sleeps* for 1 second! Can we change that to 0.1 s?

File: astropy/io/fits/tests/test_groups.py [hunk 2]
```diff
@@ -48,6 +48,7 @@ def test_open(self):
             assert ghdu.data[0].data.shape == naxes[::-1]
             assert ghdu.data[0].parnames == parameters
 
+    @pytest.mark.slow
```
Comment (mhvk) [inline astropy/io/fits/tests/test_groups.py:51]: And the same here! It sleeps for 1 second.

File: astropy/io/fits/tests/test_image.py [hunk 3]
```diff
@@ -927,6 +927,7 @@ def test_image_update_header(self):
 
     # The test below raised a `ResourceWarning: unclosed transport` exception
     # due to a bug in Python <=3.10 (cf. cpython#90476)
+    @pytest.mark.slow
```
Comment (mhvk) [inline astropy/io/fits/tests/test_image.py:930]: Here too!

File: astropy/samp/tests/test_hub.py [hunk 4]
```diff
@@ -17,6 +17,7 @@ def test_SAMPHubServer():
     SAMPHubServer(web_profile=False, mode="multiple", pool_size=1)
 
 
+@pytest.mark.slow
```
Comment (mhvk) [inline astropy/samp/tests/test_hub.py:20]: And another one that sleeps 1 sec.

File: astropy/utils/tests/test_console.py [hunk 5]
```diff
@@ -155,6 +155,7 @@ def test_progress_bar_as_generator():
     assert sum == 1225
 
 
+@pytest.mark.slow
```
Comment (mhvk) [inline astropy/utils/tests/test_console.py:158]: test_progress_bar_func.func also sleeps (though not as long)

**Comment by mhvk** [comment]
Looking through the actual tests, quite a few are slow just because they are sleeping - with 12000 tests, one really should not do even `sleep(1)` as a matter of course (but then, the FITS tests are not the newest...). I'm less sure why the utils/data tests are so slow; those tests construct whole lists of fake URLs, but by inspection it is not clear how many those actually are.

**Comment by namurphy** [comment]
Over in PlasmaPy, we've started marking tests as slow if they take longer than ∼0.25–0.5 seconds.  By doing that, and caching `.tox` between runs (https://github.com/PlasmaPy/PlasmaPy/pull/2552), we've been able to get our "skip slow" tests to finish in about a minute.  That covers ∼90% of our ∼4300 tests.  We still run the full test suite once in CI to get code coverage, and we use cron jobs our test suite to cover multiple architectures and versions of Python.  For CI, it's been incredibly helpful to get rapid feedback.

This a long way of saying...thank you for doing this! 😺

**Comment by mhvk** [comment]
I like the idea of caching - thanks for linking to that example setup!!

File: astropy/io/fits/tests/test_image.py (hunk 3 above)
Comment (neutrinoceros) [reply astropy/io/fits/tests/test_image.py:930]: I note that even cutting 90% of sleep time in this test only makes it 5 times longer than the second slowest in the module. That's about 0.2s on my machine, which is probably acceptable. Let's try !

File: astropy/utils/tests/test_console.py (hunk 5 above)
Comment (neutrinoceros) [reply astropy/utils/tests/test_console.py:158]: For this one, `sleep` time is actually not dominant, so I think it should still be marked as slow with no other changes.

File: astropy/io/fits/hdu/compressed/tests/test_compressed.py (hunk 1 above)
Comment (neutrinoceros) [reply astropy/io/fits/hdu/compressed/tests/test_compressed.py:207]: It works on my system but I'm not sure how portable this is, maybe the test would fail on slower archs/containers/VMs ? Anyway, `time.sleep(1)` is used in combinations with `pytest.mark.slow` in other similar tests, but admittedly a difference is that in this one case, `sleep` is actually responsible for most of the test's time, so let's try that.

File: astropy/io/fits/tests/test_image.py [hunk 6]
```diff
@@ -925,9 +925,6 @@ def test_image_update_header(self):
         with fits.open(self.temp("test0.fits")) as hdul:
             assert (orig_data == hdul[1].data).all()
 
-    # The test below raised a `ResourceWarning: unclosed transport` exception
-    # due to a bug in Python <=3.10 (cf. cpython#90476)
-    @pytest.mark.filterwarnings("ignore:unclosed transport <asyncio.sslproto")
```
Comment (neutrinoceros) [inline astropy/io/fits/tests/test_image.py:930]: Since this cleanup is actually orthogonal, I've opened https://github.com/astropy/astropy/pull/16183 for it.

**Comment by neutrinoceros** [comment]
Sorry I forgot about this one for a couple weeks !
In addressing @mhvk's review, I actually reverted all slow markers that I previously added in `io.fits` tests, and instead made them "faster" (or less sleepy, depending how you want to look at it).
@pllim, would you be happy with the change if I just revert all markers I added to `utils` tests too ?

**Review by mhvk — APPROVED** [review]
Looks good. I'll approve for all but `utils`, deferring to @pllim for that.

File: astropy/io/fits/tests/test_image.py (hunk 3 above)
Comment (pllim) [reply astropy/io/fits/tests/test_image.py:930]: Like I said before, I am not comfortable marking anything in `io.fits` as slow here. I will defer to @saimn .

File: astropy/utils/tests/test_data.py
Comment (pllim) [inline astropy/utils/tests/test_data.py:1]: I also not comfortable in marking `utils` tests as slow.

Comment (neutrinoceros) [reply astropy/utils/tests/test_data.py:1]: reverted

Comment (pllim) [reply astropy/utils/tests/test_data.py:1]: Thanks! Much appreciated.

**Comment by saimn** [comment]
Finally coming to this one (after catching up will the flood of notifications, taking some time off is getting harder...).
So when adding the slow mark, we used to have tests that take way more than 10 sec. (see for example the run with slow tests, https://github.com/astropy/astropy/actions/runs/8011602338/job/21885207661#step:10:1721)

Now for a run without slow tests (https://github.com/astropy/astropy/actions/runs/8011602338/job/21885207469#step:10:1717), it seems more reasonable, with only a few tests takings more than 2 sec. (on CI, which is usually much ower):
```
4.10s call     docs/timeseries/lombscarglemb.rst::lombscarglemb.rst
3.35s call     .tox/py310-test-alldeps/Lib/site-packages/astropy/coordinates/tests/test_angles.py::test_angle_multithreading
2.34s call     .tox/py310-test-alldeps/Lib/site-packages/astropy/nddata/tests/test_nduncertainty.py::test_for_leak_with_uncertainty
2.13s call     .tox/py310-test-alldeps/Lib/site-packages/astropy/coordinates/tests/test_angles.py::test_str_repr_angles_nan[input0-nan-nan deg-Angle]
2.06s setup    .tox/py310-test-alldeps/Lib/site-packages/astropy/coordinates/tests/test_angles.py::test_str_repr_angles_nan[input0-nan-nan deg-Angle]
2.00s call     .tox/py310-test-alldeps/Lib/site-packages/astropy/io/fits/tests/test_image.py::TestImageFunctions::test_open_scaled_in_update_mode
```

So unless we decide to skip all tests >1s or even less I would prefer to keep the 3 tests doing a sleep in `io.fits`.

**Comment by saimn** [comment]
And local results with a faster cpu:
```
2.85s call     docs/timeseries/lombscarglemb.rst::lombscarglemb.rst
2.01s call     astropy/io/fits/tests/test_image.py::TestImageFunctions::test_open_scaled_in_update_mode
1.90s call     astropy/nddata/tests/test_nduncertainty.py::test_for_leak_with_uncertainty
1.52s call     astropy/timeseries/periodograms/lombscargle_multiband/tests/test_lombscargle_multiband.py::test_unit_conversions[False-flexible]
1.50s call     astropy/timeseries/periodograms/lombscargle_multiband/tests/test_lombscargle_multiband.py::test_unit_conversions[True-flexible]
1.35s call     astropy/time/tests/test_precision.py::test_sidereal_lat_independent[mean]
1.26s teardown astropy/samp/tests/test_web_profile.py::TestWebProfile::test_main
1.12s teardown astropy/samp/tests/test_web_profile.py::TestWebProfile::test_web_profile
1.01s call     astropy/samp/tests/test_hub.py::test_SAMPHubServer_run
1.00s call     astropy/io/fits/hdu/compressed/tests/test_compressed.py::TestCompressedImage::test_open_comp_image_in_update_mode
1.00s call     astropy/io/fits/tests/test_groups.py::TestGroupsFunctions::test_open_groups_in_update_mode
...
================= 28290 passed, 323 skipped, 231 xfailed in 173.34s (0:02:53) ==================
```
Sure we could speedup even more by removing some tests but it seems difficult to put a limit (and what about tests that run faster but are repeated dozens of time).

File: astropy/io/fits/tests/test_image.py (hunk 3 above)
Comment (saimn) [reply astropy/io/fits/tests/test_image.py:930]: Well there is no good value here, depends on the system, disk, caching etc.
So 1s is maybe a bit too much but seems reasonable to avoid wasting time debugging issues on CI or on exotic platforms.

**Comment by neutrinoceros** [comment]
Thanks for the insight, I had no idea that 20s tests were a thing 🤯 
Let me just revert the changes to io/fits completely so we can all forget about this !
